### PR TITLE
feat(sdk): single-watch wait_for_claim_ready + faster dev port-forward polling

### PR DESCRIPTION
**Pull request by aditya-shantanu**
#### What this PR does / why we need it

Waiting for a SandboxClaim to become ready in the Python SDK currently runs **two sequential watches**: one on the claim (to learn the bound sandbox name), then a second on the sandbox (to wait for Ready). But the claim's own status already carries both the bound sandbox name and the forwarded Ready condition, so the second watch is a full watch setup/teardown of pure overhead on every claim.

Per-change summary:

- **`wait_for_claim_ready()` (sync and async):** a single watch on the claim itself that resolves on the claim's own status, anchored to the create/read `resourceVersion` so no event between the initial read and the watch start can be missed (with a 410-Gone restart path). The legacy two-watch methods are kept unchanged, so existing callers are fully compatible.
- **Dev port-forward readiness polling 500ms → 50ms:** the fixed 500ms poll dominated connect times in dev/port-forward mode; with fast-starting sandboxes it added up to ~0.5s of pure waiting per connection.
- **README latency guidance** documenting the single-watch wait and the port-forward polling behavior.

#### Performance (A/B, this change alone)

Measured 2026-07-21 on kops GCP (Kubernetes v1.35.6): 6× e2-standard-8 workers + e2-standard-16 control plane, tuned control plane/nodes, **stock upstream-main controller image** deployed (server side identical for both modes), warm pool of 60, client on an in-region VM. 50 sequential claim create → ready-wait cycles per mode; pool availability is re-checked between iterations so every cycle is a warm adoption. Legacy mode ran the upstream-main SDK; single mode ran this branch.

| claim create → ready-wait return (n=50) | legacy two-watch | single-watch (this PR) | delta |
|---|---|---|---|
| mean | 68.4ms | 54.9ms | **−13.5ms (−20%)** |
| p50 | 67.0ms | 53.5ms | −20% |
| p90 | 83.7ms | 66.0ms | −21% |
| max | 110.5ms | 82.8ms | −25% |

50/50 successes in both modes, zero failures. The saved ~13-18ms is the second watch's setup/teardown round trip — a constant client-side cost that becomes a significant share of observed ready latency exactly when the server side is fast (the same cluster's uncontended create→Ready is ~60-90ms).

Unit tests cover the new single-watch wait (ready, resourceVersion anchoring, 410/permission failure paths) for both the sync and async clients; the full package suite passes (298 tests).

#### Which issue(s) this PR fixes

Related to #574, #286

```release-note
Python SDK: new single-watch `wait_for_claim_ready()` waits for a SandboxClaim using one watch on the claim itself (previously two sequential watches); dev port-forward readiness polling interval reduced from 500ms to 50ms.
```

## Summary by CodeRabbit

* **Performance Improvements**
  * Reduced sandbox startup latency by switching `create_sandbox` to a single, watch-based claim readiness flow.
  * Improved local-tunnel readiness detection by polling every 50ms.

* **Reliability**
  * Enhanced claim readiness handling to resume from the last seen event, recover from Kubernetes watch compaction (HTTP 410), and correctly handle warm vs cold readiness timing.
  * Added `SandboxClaimFailedError` for terminal claim failures.

* **Documentation**
  * Added a new Python client section explaining startup latency, warm/cold behavior, timeouts, and local-tunnel overhead.

* **Tests**
  * Expanded unit coverage for readiness edge cases and watch restart/410 recovery.

### CODE REVIEW

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py [hunk 1]
```diff
@@ -91,72 +91,140 @@ def create_sandbox_claim(
...
+                    timeout_seconds=remaining
+                )
+                for event in event_iter:
+                    if event is None:
+                        continue
+                    if event["type"] == "DELETED":
                         w.stop()
-                        return name
+                        raise SandboxMetadataError(
+                            f"SandboxClaim '{claim_name}' was deleted while resolving sandbox name")
```
[the remark this answers is not in the document]
Comment (aditya-shantanu) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py]: Done in 159e292 — the shared loop now uses the goal-aware message: "deleted while waiting for {claim readiness|sandbox name}" (existing test updated to match).

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py [hunk 2]
```diff
@@ -91,72 +91,140 @@ def create_sandbox_claim(
...
+                                raise SandboxTemplateNotFoundError(
+                                    f"SandboxTemplate requested does not exist: {cond.get('message', 'Template not found')}"
+                                )
+                            elif cond.get('reason') == 'WarmPoolNotFound':
+                                w.stop()
+                                raise SandboxWarmPoolNotFoundError(
+                                    f"SandboxWarmPool requested does not exist: {cond.get('message', 'WarmPool not found')}"
+                                )
+                            if cond.get('type') == 'Ready' and cond.get('status') == 'True':
+                                ready = True
```
[the remark this answers is not in the document]
Comment (aditya-shantanu) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py:223]: Done in 159e292 — added TERMINAL_CLAIM_READY_REASONS (InvalidMetadata, EnvVarsInjectionRejected, VolumeClaimTemplatesError, ClaimExpired, SandboxExpired) shared via constants.py; both sync and async watch loops now raise the new SandboxClaimFailedError immediately on those. I deliberately kept ReconcilerError out of the terminal set: it's the controller's generic fallback for retried errors (e.g. transient conflicts), so failing the SDK fast on it would turn recoverable blips into client errors — covered by a new keeps-waiting test alongside the fail-fast test.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py [hunk 3]
```diff
@@ -154,8 +201,14 @@ async def resolve_sandbox_name(self, claim_name: str, namespace: str, timeout: i
                         )
```
[the remark this answers is not in the document]
Comment (aditya-shantanu) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py:206]: Done in 159e292 — async variant uses the same goal-aware message.

**Comment by aditya-shantanu** [comment]
SDK health validation on a live tuned 6-node kops cluster (stock upstream-main controller deployed): 150 sequential `create → wait_for_claim_ready() → delete` cycles with this branch's SDK — **150/150 succeeded, 0 failures** (mean 60ms, p50 52.1ms, p90 89.9ms, max 162ms; every cycle a warm adoption), plus a 30-cycle legacy-path sanity run (30/30) confirming the retained two-watch methods keep working. resourceVersion-anchored watch behavior held across all cycles (no missed events, no stuck waits).

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py [hunk 4]
```diff
@@ -110,29 +111,80 @@ async def create_sandbox_claim(
...
+        with UNSET resourceVersion forces a quorum etcd read to establish
+        initial state on every wait, which at high claim rates is pure
+        apiserver/etcd load on the latency path. If the supplied version has
+        been compacted away (410 Gone), the watch transparently restarts from
+        ``"0"`` (current state from the watch cache).
         """
         await self._ensure_initialized()
 
+        goal = "claim readiness" if require_ready else "sandbox name"
+        # Keep the legacy resolve-path message byte-identical to the original
```
Comment (SHRUTI6991) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py]: Can we remove this comment? Seems like this is just AI generated.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py [hunk 5]
```diff
@@ -110,29 +111,80 @@ async def create_sandbox_claim(
...
+        ``"0"`` — so the apiserver serves it from the watch cache. A watch
+        with UNSET resourceVersion forces a quorum etcd read to establish
+        initial state on every wait, which at high claim rates is pure
+        apiserver/etcd load on the latency path. If the supplied version has
+        been compacted away (410 Gone), the watch transparently restarts from
+        ``"0"`` (current state from the watch cache).
         """
         await self._ensure_initialized()
 
+        goal = "claim readiness" if require_ready else "sandbox name"
```
Comment (SHRUTI6991) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py]: why are we naming this goal?

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_sandbox_client.py [hunk 6]
```diff
@@ -441,8 +444,15 @@ async def _create_claim(
...
         )
 
+    @async_trace_span("wait_for_claim_ready")
+    async def _wait_for_claim_ready(self, claim_name: str, namespace: str, timeout: int, resource_version: str | None = None) -> str:
+        """Waits for the SandboxClaim to be bound and Ready, returning the sandbox name."""
+        return await self.k8s_helper.wait_for_claim_ready(claim_name, namespace, timeout, resource_version=resource_version)
+
     @async_trace_span("wait_for_sandbox_ready")
     async def _wait_for_sandbox_ready(self, sandbox_id: str, namespace: str, timeout: int):
+        """Retained for API compatibility; ``create_sandbox`` now uses the
```
Comment (SHRUTI6991) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_sandbox_client.py]: Unless we plan to support both paths, we should just remove these comments.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py (hunk 4 above)
Comment (aditya-shantanu) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py]: Fixed in 57a95f0 — the comment did encode a real constraint (the unit tests assert on the exact resolve-path message), so I replaced it with a one-line note stating just that; also swept the identical comment in the sync k8s_helper.py.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py (hunk 5 above)
Comment (aditya-shantanu) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py]: Fixed in 57a95f0 — renamed to wait_target (it holds the human-readable description of what the watch is waiting for, used only in the log/timeout messages); renamed in the sync helper too.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_sandbox_client.py (hunk 6 above)
Comment (aditya-shantanu) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_sandbox_client.py]: Fixed in 57a95f0 — dropped the dual-path commentary; the docstring now just says what the method does (the method itself stays since the unit tests still exercise it). Also swept the identical docstring in the sync sandbox_client.py.

**Comment by SHRUTI6991** [comment]
nice change, thank you.

/lgtm
